DEVELOPING.md

```diff
@@ -1,75 +1,210 @@
+git remote
 ```
+If you have a remote named `origin`, you can see the repository it is pointing to with:
+```bash
+git remote get-url origin
+```
+You want to have at least two remotes, one pointing to upstream and another pointing to your user fork.
+Add an `upstream` remote with this command:
+```bash
+git remote add upstream git@github.com:tlaplus/tlaplus.git
+```
+If you cloned your repo from your user fork, you can also rename `origin` to `user` or something similarly descriptive as follows:
+```bash
+git remote rename origin user
+```
+If you want to get the latest changes from upstream, you can then run:
+```bash
+git checkout master
+git pull upstream master
+```
+Then, either merge or rebase on those changes to get them into your local development branches.
+
+### Modifying past commits
+
+When you're working with others, you don't want to subject them to your raw git commit history.
+Often you accumulate commits that are only periodic check-ins to avoid losing work instead of logically-contained sets of changes.
+Git's `rebase` command is much-feared but powerful, and you should become very familiar with it.
+If you are developing a feature in a branch, you can enter rebase mode by running:
+```bash
+git rebase -i master
+```
+This will take you to a text editor where you can command git to do all kinds of things with your commits!
+You can:
+ * Reorder them
+ * Squash several of them into a single commit
+ * Select them to be manually edited/amended
+
+Then, git will run through all your commits sequentially and execute whatever manipulations you indicated.
+Squashing is very useful for combining those periodic check-in commits into something meaningful.
+Editing is useful for responding to PR review comments; most people know about using `git --amend` to apply changes to the latest commit, but using the `rebase` edit option lets you apply changes to commits in the past!
 
-The `tlatools.jar` will be created in the `org.lamport.tlatools/dist` directory.
+### Splitting up commits and PRs
 
-Test that the build was successful by using an example model:
+While rebasing, you can also use editing to split a single commit up into multiple commits.
+This is fairly straightforward when the changes are in different files, but if you want some changes within a single file to go in one commit and some to go in another commit, you can accomplish this interactively with:
+```bash
+git add --patch <filename>
+```
+Sometimes you might want to split your commits across multiple branches so they can be in separate PRs.
+For this, switch to a new branch then use:
 ```bash
-$ java -jar dist/tla2tools.jar ./test-model/pcal/Bakery.tla
+git cherry-pick <commit hash>
 ```
-To run the unit tests, abbreviate the above `ant` command:
+This will add the specified commit to the head of your new branch.
+You can then rebase your original feature branch to remove that commit.
+
+### Developer Certificate of Origin (DCO) signing
```

Comment:
Let's not call this "signing" but "sign-off", because the former sounds too much like GPG signatures...

/cc @Calvin-L